### ISSUE TITLE
Useless use of string eq in void context at t/lib/TestBridge.pm line 351

### ISSUE THREAD

**Issue by mauke** [Perl-Toolchain-Gang/YAML-Tiny#66]
The test code in TestBridge.pm has a precedence bug: It uses `or` where it should use `||`, leading to warnings and (more importantly) tests that always pass because the test description is used as a boolean (and always true).

The bug was introduced in commit 128b71f24bf416.

Also reported as https://github.com/Perl-Toolchain-Gang/CPAN-Meta-YAML/issues/6.

```
$ prove -l
t/01_api.t ............ Useless use of string eq in void context at t/lib/TestBridge.pm line 351.
Useless use of string eq in void context at t/lib/TestBridge.pm line 356.
t/01_api.t ............ ok   
t/01_compile.t ........ 1/? Useless use of string eq in void context at t/lib/TestBridge.pm line 351.
Useless use of string eq in void context at t/lib/TestBridge.pm line 356.
t/01_compile.t ........ ok   
t/10_read.t ........... Useless use of string eq in void context at t/lib/TestBridge.pm line 351.
Useless use of string eq in void context at t/lib/TestBridge.pm line 356.
t/10_read.t ........... ok   
t/11_read_string.t .... Useless use of string eq in void context at t/lib/TestBridge.pm line 351.
Useless use of string eq in void context at t/lib/TestBridge.pm line 356.
t/11_read_string.t .... ok   
t/12_write.t .......... Useless use of string eq in void context at t/lib/TestBridge.pm line 351.
Useless use of string eq in void context at t/lib/TestBridge.pm line 356.
t/12_write.t .......... ok   
t/13_write_string.t ... Useless use of string eq in void context at t/lib/TestBridge.pm line 351.
Useless use of string eq in void context at t/lib/TestBridge.pm line 356.
t/13_write_string.t ... ok   
t/20_subclass.t ....... ok   
t/21_yamlpm_compat.t .. Useless use of string eq in void context at t/lib/TestBridge.pm line 351.
Useless use of string eq in void context at t/lib/TestBridge.pm line 356.
t/21_yamlpm_compat.t .. ok   
t/30_yaml_spec_tml.t .. Useless use of string eq in void context at t/lib/TestBridge.pm line 351.
Useless use of string eq in void context at t/lib/TestBridge.pm line 356.
# using JSON backend: Cpanel::JSON::XS 4.38
t/30_yaml_spec_tml.t .. ok   
t/31_local_tml.t ...... Useless use of string eq in void context at t/lib/TestBridge.pm line 351.
Useless use of string eq in void context at t/lib/TestBridge.pm line 356.
t/31_local_tml.t ...... ok    
t/32_world_tml.t ...... Useless use of string eq in void context at t/lib/TestBridge.pm line 351.
Useless use of string eq in void context at t/lib/TestBridge.pm line 356.
t/32_world_tml.t ...... ok    
t/86_fail.t ........... ok   
All tests successful.
Files=12, Tests=57,  2 wallclock secs ( 0.13 usr  0.02 sys +  1.21 cusr  0.16 csys =  1.52 CPU)
Result: PASS
```

**Comment by mauke** [comment]
Quoting from PR #60 by @azertyfun:

> It used to be that all "special" YAML keywords (null, true, false, yes, no, etc.) would be quoted to avoid `Dump("yes") eq "---\nyes"` since YAML `yes` is a boolean, not a string.
> 
> For some reason, the commit silently changed that behavior to /only/ escape `null`, `true`, and `false`. I cannot find any rationale for it, as `yes` is equivalent to `true` so why would only one of them be quoted?

The rationale is that yes/no/y/n are not YAML keywords and should not be interpreted as booleans (at least by default). There is an explanation by @ingydotnet  in commit 438b30291e7144.

PR #60 changed the test code from `is $dump, $yaml` to `ok($dump eq $yaml or "'$dump'" eq $yaml`, but that doesn't make any sense: At this point, `$yaml` contains something like `"y: y\n"` and `$dump` something like `"'y': 'y'\n"`. You can't make those match by simply wrapping `' '` around `$dump`. The only reason the tests didn't fail immediately was the use of low-precedence `or`, which effectively disables the tests and always returns true.

I think #60 should be reverted.

**Comment by mauke** [comment]
The spec file (`t/tml-spec/unicode.tml`) that controls the relevant test (`t/30_yaml_spec_tml.t`) contains this:
```
# Some implementations think N means false. This should not be the case in a
# default schema. No quotes.
#
# NOTE:
#   http://yaml.org/type/bool.html suggests that many simple strings should be
#   loaded as boolean, but this is an outdated concept. Currently, only the
#   words true/false/null (lower case) should be loaded specially (not as
#   strings).  This may become even more restrictive in the future. ie Only
#   true/false/null in a flow context.
=== Code point 78
--- code: 78
--- yaml
N: N


# Some implementations think Y means true. This should not be the case in a
# default schema. No quotes.
=== Code point 89
--- code: 89
--- yaml
Y: Y
```

In other words, the "regression" back in 2014 was very much intentional (and it has been this way for over a decade).

If we want to change the behavior of YAML::Tiny now, we should edit the tml file and not fiddle with the string comparison logic.

**Comment by karenetheridge** [comment]
Interesting; I always thought `yes` was to be treated as a boolean as well!

> I think https://github.com/Perl-Toolchain-Gang/YAML-Tiny/pull/60 should be reverted.

agreed.

**Comment by karenetheridge** [comment]
reverted in 1.76.

**Comment by perlpunk** [comment]
FYI: Please see https://perlpunk.github.io/yaml-test-schema/schemas.html for how booleans and other special values should be resolved in YAML 1.1 and 1.2.
YAML::PP is currently the only perl module implementing this according to the spec. YAML::XS will follow soon when I find the time to finish my work from november on that.
